crates/biome_cli/src/reporter/gitlab.rs

```diff
@@ -0,0 +1,243 @@
+                }
+
+                self.console.log(markup!({ content }));
+            }
+        }
+        self.console.log(markup!("]"));
+        Ok(())
+    }
+}
+
+pub(crate) struct GitLabDiagnosticBuilder {
+    /// The root of the Git repository.
+    /// Required for relativization of the reported absolute paths.
+    repository_root: Option<PathBuf>,
+    /// A set of fingerprints (unique identifiers) to prevent collisions.
+    fingerprints: HashSet<u64>,
+}
+
+impl GitLabDiagnosticBuilder {
+    pub fn new(repository_root: Option<PathBuf>) -> Self {
+        Self {
+            repository_root,
+            fingerprints: HashSet::new(),
+        }
+    }
+
+    /// Turns a biome diagnostic into one in the GitLab Code Quality report format.
+    pub fn gitlab_diagnostic(
+        &mut self,
+        value: biome_diagnostics::error::Error,
+    ) -> GitLabDiagnostic {
+        let check_name = self.check_name(&value).unwrap_or_default();
+        let path = self.path(&value).unwrap_or_default();
+        let line = self.line(&value).map_or(1, |line| max(line, 1));
+        let source_code = self.source_code(&value).unwrap_or_default();
+        let fingerprint = self
+            .ensure_fingerprint_uniqueness(self.fingerprint(&check_name, &path, &source_code), 0)
+            .to_string();
+
+        GitLabDiagnostic {
+            severity: GitLabSeverity::from(value.severity()),
+            description: PrintDescription(&value).to_string(),
+            check_name,
+            fingerprint,
+            location: Location {
+                path,
+                lines: Lines { begin: line },
+            },
+        }
+    }
+
+    /// Extracts the name of the category as the check name.
+    fn check_name(&self, value: &biome_diagnostics::error::Error) -> Option<String> {
+        let category = value.category()?;
+        Some(category.name().to_string())
+    }
+
+    /// Extracts the source code generating the diagnostic.
+    fn source_code(&self, value: &biome_diagnostics::error::Error) -> Option<String> {
+        let location = value.location();
+        let source_code = location.source_code?;
+        let span = location.span?;
+
+        Some(source_code.text[span].to_string())
+    }
+
+    /// Generates a fingerprint for a diagnostic.
+    fn fingerprint(&self, check_name: &String, path: &String, code: &String) -> u64 {
+        let mut hasher = DefaultHasher::new();
+
+        // Including the source code in our hash leads to more stable
+        // fingerprints. If you instead rely on e.g. the line number and change
+        // the first line of a file, all of its fingerprint would change.
+        code.hash(&mut hasher);
+        check_name.hash(&mut hasher);
+        path.hash(&mut hasher);
+
+        hasher.finish()
+    }
+
+    /// Enforces uniqueness of generated fingerprints in the context of a
+    /// single report.
+    fn ensure_fingerprint_uniqueness(&mut self, fingerprint: u64, salt: u64) -> u64 {
```

Comment:
How does this method enforces the uniqueness? This isn't explained. Plus, I was expecting `Result`, so maybe better explanation can remove my doubts

Reply:
It only needs to be unique for a specific run, So we can get away with just "trying again", until we obtain a hash that has not been seen before. Looking at it again, we can probably omit the salt, it seems unnecessary.

Here is a playground roughly demonstrating the behavior: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=052a51816c57fbc363d0e5104cfdecd4

The other two implementations mentioned in the PR also follow similar approaches:
- `gitlab-formatter-eslint` https://gitlab.com/gitlab-formatters/eslint-formatter-gitlab/-/blob/main/src/functions.ts?ref_type=heads#L19-25 rehashes with a random string
- `ruff` https://github.com/astral-sh/ruff/blob/main/crates/ruff_linter/src/message/gitlab.rs#L86-L91 is what I used as an orientation. They also re-hash the duplicate until it is unique.